scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1095,18 +1095,20 @@ final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) ext
       // there is no actual merging here, no need to rename:
       merged.head
     } else {
-      merged.reduce[Pipe] {
-        case (left, right) =>
-          // special handling for cases where one side of the hashjoin is merged
-          // with the hashjoin result. Cascading no longer allows it,
-          // so we add a checkpoint to the join result as a workaround
-          if (RichPipe.isHashJoinedWithPipe(left, right))
-            new Merge(RichPipe.assignName(new Checkpoint(left)), RichPipe.assignName(right))
-          else if (RichPipe.isHashJoinedWithPipe(right, left))
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(new Checkpoint(right)))
-          else
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(right))
-      }
+      val preparedSources = merged.map(p => {
+        merged
+          .filterNot(_ == p)
+          .find(other => RichPipe.isHashJoinedWithPipe(p, other)) match {
+            case Some(_) =>
+              // special handling for cases where one side of the hashjoin is merged
+              // with the hashjoin result. Cascading no longer allows it,
+              //so we add a checkpoint to the join result as a workaround
+              new Checkpoint(p)
```

Comment:
I think one issue here might be that this can add more Checkpoints compared to the earlier version that uses `reduce`. (If we've already added a checkpoint to `left`, then we do not need another per `isHashJoinedWithPipe`)